File: DouYu/app/src/main/java/cn/edu/app/douyu/network/DoyuApiClient.java
package cn.edu.app.douyu.network;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.GsonBuilder;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import cn.edu.app.douyu.BuildConfig;
import cn.edu.app.douyu.data.DoyuRepository;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * API 客户端工厂：配置 Retrofit、OkHttp、认证拦截器并创建 Repository。
 */
public final class DoyuApiClient {
    private DoyuApiClient() {
    }

    // 创建业务 API 客户端；SessionInterceptor 会自动附加登录 token。
    public static DoyuApi create(Context context) {
        SharedPreferences preferences = context.getSharedPreferences("doyu_session", Context.MODE_PRIVATE);
        OkHttpClient client = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(20, TimeUnit.SECONDS)
                .addInterceptor(new SessionInterceptor(preferences))
                .build();

        // Retrofit 负责把 DoyuApi 接口声明转换成真实 HTTP 请求，Gson 负责 JSON 映射。
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BuildConfig.API_BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(new GsonBuilder().create()))
                .build();

        return retrofit.create(DoyuApi.class);
    }

    // 页面统一拿 Repository，不直接持有 Retrofit 接口，方便集中处理上传和返回值。
    public static DoyuRepository createRepository(Context context) {
        return new DoyuRepository(create(context), createUploadClient(), BuildConfig.API_BASE_URL);
    }

    // 上传图片使用独立 OkHttpClient，因为 presign 后的 PUT 请求不走普通 Retrofit API。
    static OkHttpClient createUploadClient() {
        return new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    /**
     * 会话拦截器：给请求加 X-Request-Id 和 Bearer token，并兼容社区公开读接口的匿名访问。
     */
    static final class SessionInterceptor implements Interceptor {
        private final SharedPreferences preferences;

        SessionInterceptor(SharedPreferences preferences) {
            this.preferences = preferences;
        }

        @Override
        public Response intercept(Chain chain) throws IOException {
            Request original = chain.request();
            String token = preferences.getString("accessToken", "");
            Request request = addBaseHeaders(original, token);
            Response response = chain.proceed(request);
            // 公开社区读接口如果带过期 token 返回 401，就清掉 token 后匿名重试一次。
            if (response.code() == 401 && !token.isEmpty() && isAnonymousCommunityRead(original.method(), original.url().encodedPath())) {
                response.close();
                preferences.edit().remove("accessToken").apply();
                return chain.proceed(addBaseHeaders(original, ""));
            }
            return response;
        }

        // 统一添加请求追踪头；登录后再补 Authorization。
        private Request addBaseHeaders(Request original, String token) {
            Request.Builder builder = original.newBuilder()
                    .header("X-Request-Id", "android-java-xml");
            if (!token.isEmpty()) {
                builder.header("Authorization", "Bearer " + token);
            } else {
                builder.removeHeader("Authorization");
            }
            return builder.build();
        }

        // 这些 GET 接口允许未登录访问，过期 token 不应阻断首屏内容。
        static boolean isAnonymousCommunityRead(String method, String path) {
            if (!"GET".equalsIgnoreCase(method)) {
                return false;
            }
            return "/api/v1/posts/feed".equals(path)
                    || path.matches("/api/v1/topics(/[^/]+/posts)?")
                    || path.matches("/api/v1/posts/[^/]+/comments")
                    || path.matches("/api/v1/posts/[^/]+");
        }
    }
}

File: DouYu/app/src/main/java/cn/edu/app/douyu/core/network/DoyuApiClient.kt
package cn.edu.app.douyu.core.network

import cn.edu.app.douyu.core.model.RefreshTokenRequest
import cn.edu.app.douyu.core.model.TokenPair
import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import kotlinx.coroutines.runBlocking
import okhttp3.Authenticator
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.Route
import retrofit2.Retrofit
import java.util.UUID
import java.util.concurrent.TimeUnit

class AuthInterceptor(
    private val tokenProvider: AccessTokenProvider
) : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val token = tokenProvider.accessToken()
        val authenticatedRequest = if (!token.isNullOrBlank()) {
            request.newBuilder()
                .header(ApiHeaders.AUTHORIZATION, "Bearer $token")
                .build()
        } else {
            request
        }

        return chain.proceed(authenticatedRequest)
    }
}

class IdempotencyKeyInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val needsKey = request.method in IDEMPOTENT_SUBMISSION_METHODS &&
            request.header(ApiHeaders.IDEMPOTENCY_KEY).isNullOrBlank()
        val next = if (needsKey) {
            request.newBuilder()
                .header(ApiHeaders.IDEMPOTENCY_KEY, UUID.randomUUID().toString())
                .build()
        } else {
            request
        }
        return chain.proceed(next)
    }

    private companion object {
        val IDEMPOTENT_SUBMISSION_METHODS = setOf("POST", "PATCH", "DELETE")
    }
}

class TokenRefreshAuthenticator(
    private val refreshTokenProvider: RefreshTokenProvider,
    private val onTokenRefreshed: suspend (TokenPair) -> Unit,
    private val onRefreshFailed: suspend () -> Unit,
    private val refreshServiceFactory: () -> AuthApi
) : Authenticator {
    override fun authenticate(route: Route?, response: Response): Request? {
        if (responseCount(response) >= 2) return null
        if (response.request.url.encodedPath.endsWith("/api/v1/auth/refresh")) return null

        val refreshToken = refreshTokenProvider.refreshToken().orEmpty()
        if (refreshToken.isBlank()) {
            runBlocking { onRefreshFailed() }
            return null
        }

        val refreshed = runBlocking {
            runCatching {
                refreshServiceFactory().refresh(RefreshTokenRequest(refreshToken))
            }.getOrNull()
        }

        val tokenPair = refreshed?.data
        return if (refreshed?.isOk == true && tokenPair != null) {
            runBlocking { onTokenRefreshed(tokenPair) }
            response.request.newBuilder()
                .header(ApiHeaders.AUTHORIZATION, "Bearer ${tokenPair.accessToken}")
                .build()
        } else {
            runBlocking { onRefreshFailed() }
            null
        }
    }

    private fun responseCount(response: Response): Int {
        var count = 1
        var prior = response.priorResponse
        while (prior != null) {
            count++
            prior = prior.priorResponse
        }
        return count
    }
}

class DoyuApiClient(
    baseUrl: String,
    tokenProvider: AccessTokenProvider,
    refreshTokenProvider: RefreshTokenProvider = object : RefreshTokenProvider {
        override fun refreshToken(): String? = null
    },
    private val onTokenRefreshed: suspend (TokenPair) -> Unit = {},
    private val onAuthExpired: suspend () -> Unit = {},
    okHttpClientBuilder: OkHttpClient.Builder = OkHttpClient.Builder()
) {
    private val refreshRetrofit: Retrofit
    private val retrofit: Retrofit

    init {
        val refreshClient = OkHttpClient.Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build()

        refreshRetrofit = createRetrofit(baseUrl, refreshClient)

        val client = okHttpClientBuilder
            .addInterceptor(AuthInterceptor(tokenProvider))
            .addInterceptor(IdempotencyKeyInterceptor())
            .authenticator(
                TokenRefreshAuthenticator(
                    refreshTokenProvider = refreshTokenProvider,
                    onTokenRefreshed = onTokenRefreshed,
                    onRefreshFailed = onAuthExpired,
                    refreshServiceFactory = { refreshRetrofit.create(AuthApi::class.java) }
                )
            )
            .connectTimeout(5, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build()

        retrofit = createRetrofit(baseUrl, client)
    }

    val authApi: AuthApi = retrofit.create(AuthApi::class.java)
    val userApi: UserApi = retrofit.create(UserApi::class.java)
    val uploadApi: UploadApi = retrofit.create(UploadApi::class.java)
    val communityApi: CommunityApi = retrofit.create(CommunityApi::class.java)
    val patternApi: PatternApi = retrofit.create(PatternApi::class.java)
    val productApi: ProductApi = retrofit.create(ProductApi::class.java)
    val cartApi: CartApi = retrofit.create(CartApi::class.java)
    val orderApi: OrderApi = retrofit.create(OrderApi::class.java)
    val paymentApi: PaymentApi = retrofit.create(PaymentApi::class.java)
    val messageApi: MessageApi = retrofit.create(MessageApi::class.java)
    val rewardApi: RewardApi = retrofit.create(RewardApi::class.java)

    private fun createRetrofit(baseUrl: String, client: OkHttpClient): Retrofit =
        Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(client)
            .addConverterFactory(DoyuJson.asConverterFactory(DoyuJsonMediaType))
            .build()
}
